### DISCUSSION TITLE
How do I type a parent-walk?

### DISCUSSION THREAD

**Discussion by cjw296** [python/typing#1830] [Q&A] [answered]
Given this code:

```python
T = TypeVar('T')

@dataclass()
class Element(Generic[T]):
    value: T
    parent: 'Element[Any] | None' = None

...

            breadcrumbs = [self.current_element]
            current = self.current_element
            while current := current.parent:
                breadcrumbs.append(current)
```

`self.current_element` is always of type `Element[Any]`.

The above code causes mypy to complain:

```
error: Item "None" of "Element[Any] | None" has no attribute "parent"  [union-attr]
```

How do I add annotations such that mypy can understand what's going on here?

**Comment by hauntsaninja** [comment]
That seems like a bug in mypy, I would type ignore.

**Reply by cjw296** [reply]
I've type ignored for now, but I have a feeling that the mypy folks would disagree. `parent` is `Element[Any] | None`, so from a static analysis perspective, it appears to be able to be None in that expression, even though in reality it never can be...

**Reply by JelleZijlstra** [reply]
Shantanu is a mypy maintainer, so at least one of them would not disagree :)

I agree that mypy should understand that you're narrowing the type of `current` here so it cannot be None.

**Reply by cjw296** [reply]
Hah! Well, happy to hear that then... do I need to submit a mypy bug or has it already happened?

**Reply by cjw296** [reply]
Any news on this?

**Reply by JukkaL** [reply]
Creating a mypy issue could help move this forward.

**Reply by cjw296** [reply, marked as answer]
https://github.com/python/mypy/issues/18608 which was marked as a duplicate of https://github.com/python/mypy/issues/2008
